rasa/core/policies/intent_ted_policy.py

```diff
@@ -0,0 +1,857 @@
+        priority: int = DEFAULT_POLICY_PRIORITY,
+        max_history: Optional[int] = None,
+        model: Optional[RasaModel] = None,
+        fake_features: Optional[Dict[Text, List["Features"]]] = None,
+        intent_thresholds: Dict[int, float] = None,
+        all_labels: List[Text] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Declare instance variables with default values."""
+        super().__init__(
+            featurizer, priority, max_history, model, fake_features, **kwargs
+        )
+
+        self._all_labels = all_labels
+        self.intent_thresholds = intent_thresholds
+        self.ignore_intent_list = self.config["intents_to_ignore"]
+        self.ignore_retrieval_intents = self.config["ignore_retrieval_intents"]
+
+        # Set all invalid configuration parameters
+        self.config[ENTITY_RECOGNITION] = False
+
+    @staticmethod
+    def supported_data() -> SupportedData:
+        return SupportedData.ML_AND_RULE_DATA
+
+    @staticmethod
+    def _standard_featurizer(max_history: Optional[int] = None) -> TrackerFeaturizer:
+        return IntentMaxHistoryFeaturizer(
+            IntentTokenizerSingleStateFeaturizer(), max_history=max_history
+        )
+
+    def _create_label_data(
+        self, domain: Domain, interpreter: NaturalLanguageInterpreter
+    ) -> Tuple[RasaModelData, List[Dict[Text, List["Features"]]]]:
+        # encode all label_ids with policies' featurizer
+        state_featurizer = self.featurizer.state_featurizer
+
+        # TODO: Change TED also to do this before label data is created.
+        #  Currently call to this is made inside featurize_trackers which masks
+        #  the problem that labels may also need to be featurized from states prepared from domain.
+        state_featurizer.prepare_for_training(domain, interpreter)
+
+        # Add an extra PAD to the vocabulary of labels.
+        # Needed because labels can be multi-label and
+        # padded to get to the same size
+        self._all_labels = ["PAD"] + domain.intents
+
+        encoded_all_labels = state_featurizer.encode_all_labels(
+            domain, interpreter, self._all_labels
+        )
+
+        attribute_data, _ = convert_to_data_format(encoded_all_labels)
+
+        label_data = RasaModelData()
+        label_data.add_data(attribute_data, key_prefix=f"{LABEL_KEY}_")
+        label_data.add_lengths(
+            f"{LABEL}_{INTENT}", SEQUENCE_LENGTH, f"{LABEL}_{INTENT}", SEQUENCE
+        )
+
+        label_ids = np.arange(len(self._all_labels))
+        label_data.add_features(
+            LABEL_KEY,
+            LABEL_SUB_KEY,
+            [FeatureArray(np.expand_dims(label_ids, -1), number_of_dimensions=2)],
+        )
+
+        return label_data, encoded_all_labels
+
+    @staticmethod
+    def _get_label_key_to_ids_map(labels: List[Text]) -> Dict[Text, int]:
+        return {label: index for index, label in enumerate(labels)}
+
+    def train(
```

Comment:
why do we need to override this?

Reply:
The threshold calculation is done at the end of this method. But, I think I'll push it to a separate method called `post_training_tasks()` or something which would be inside `TEDPolicy` and overridden by `IntentTEDPolicy`